### DISCUSSION TITLE
Need to accelerate this project - becoming urgent.

### DISCUSSION THREAD

**Discussion by mburns08109** [twinbasic/twinbasic#740] [General]
This request is not about the technical features of this effort in any way.
It is about the business aspect of this project.

With the recent screw-ups that MS has made (in the name of security, generally) they have threatened the practicality of continuing to use and support Microsoft Access as a reasonable tool for developing software solutions for business users. (see discussion here for more on this point: <https://www.devhut.net/does-microsoft-access-remain-a-good-choice-as-a-development-platform/> ) - once MS Access MVPs start asking questions like that publicly; it is something to give one pause for contemplation on both the question itself and the answers to it.

This will create a new and briefly-open temporal window of opportunity to step into the void MS is creating by blundering and placing new and deliberate obstacles into the path of the MS-Access solution-scape.

Thus, the sooner that twinBasic can be crafted into a practical alternative solution to the "heavier" software development alternatives in the marketplace, the sooner we can begin marketing tB into that market and problem space, and the better off this project, its supporters, and the general business market will be.

So my question is: presuming my thesis above is correct, what can we do to significantly accelerate this project into a version 1.0 release level? I would posit that one aspect of any answer to that question will be Wayne's desire and ability to relax his control of this effort to one degree or another. I am in no way seeking to wrest his credit or control of the direction of this, but if additional development hands and minds are required, then some degree of that is a direct consequence of doing so. 

So, I'm asking this question now due mostly to the circumstances of the time - if only to stimulate the "next level" thinking needed to possibly move this from a hobbiest-level undertaking (not meaning to be derogatory with that term) to a seriously professional effort to get a solution into the marketplace in a reasonable timeframe.

**Comment by mansellan** [comment]
Interesting. I had no idea Access was so neglected of late...

Maybe twinBASIC should try to facilitate use with the free [SQL Server Express](https://en.wikipedia.org/wiki/SQL_Server_Express)? Although it's not quite as easy to share between multiple users (needs a server), it is highly unlikely to corrupt data on a whim either (because it has a server). It can also be embedded (LocalDB) for single-user serverless use.

Obviously you don't get all the RAD tooling that Access also provides, but you _can_ use a cut-down version of Reporting Services for reports, and either SQL Server Management Studio or VS Code extensions for editing queries.

**Reply by bclothier** [reply]
I think one major prerequisite for tB to be a viable replacement for Access is to have a robust support for data binding, which from the current roadmap is still some way off. Mind, Access' data binding is quite different from VB6 data binding, and I think there are many opportunities to make it better but that's putting the cart in front of the horse. 

I agree that Microsoft has problems with understanding that Access doesn't really fit in the "document ecosystem" that the Office fosters on the rest of Office applications and it's causing a lot of pains for those who has developed applications (not documents!) based on Access.

**Comment by WaynePhillipsEA** [comment]
Thanks for opening this discussion.  Having other business ties with Microsoft Access, I am acutely aware of this ongoing and developing situation.   With new obstacles being put before using VBA in Office projects soon, coupled of course all of the update problems that have tormented Access developers over the last few years, I personally feel there's no longer a strong argument for using Microsoft Access for any **_new_** development work.  I also do agree that this situation could potentially become a good opportunity for tB...

When I first launched the initial preview of tB, I had high hopes that I would by now be able to employ someone to help with the development workload for tB, but unfortunately uptake in the pre-order licences is just not where we need it to be for that.  I do know there is an awful lot of interest in tB, and that helps strengthen my belief that the effort I'm putting into this project will not be wasted in the long-term.  It simply takes time to establish a presence in this marketplace, and I agree that until v1 is ready this situation is unlikely to change substantially.

I do realise that I haven't really answered your questions at this stage, I just wanted to be honest and bring you guys up to speed from a business perspective.   Just to be clear... twinBASIC development continues daily, and there's no planned change in this regard(!).  I strongly believe that tB is a viable, and needed product.  I am still 100% committed to making this project work.   As always, I must give massive thanks to all of our supporters that help both financially and practically (with the bug/issue reports and other discussions);  you guys are amazing.

**Reply by mburns08109** [reply]
Perhaps part of the "income problem" is also a messaging problem?

A bold statement of intent to make the tB infrastructure into not only a viable (and being updated) VB6 replacement but also as _a path forward for MS-Access VBA projects_ may add sufficient interest (at least amongst Access Developers)? 

That said, getting into data binding and your up-sleeve tricks for VBA...a year from now may not be "soon enough" to strike within the market opportunity now opening. (just my $0.02)

...any chance to work with some universities and get some intern-level help going?

**Comment by DinyaZ** [comment]
What if you try to contact the RAD Basic developer guy? His goal is the same, to develop a modern, 100% VB6 compatible language. I think, your compiler and "under the hood" details are much more mature, but his project has also an important advantage: he develops a clear, simple, and user friendly GUI, editor, form designer interface, which is easier to use, and resembles to the VB6. I think, the biggest disadvantage of TwinBASIC now is the VSCode reliance.
(RAD Basic is a typical "one-man show", it improves very slowly, I dont think it can be a real, working tool in our life.)
.
But a possible cooperation and the combination of the advantages of both projects can give the best and fastest result.....!? One working tool is better then two half.....?

**Comment by dmrvb** [comment]
That's a nice idea, a little collaboration would be good, but I doubt it will happen. The history of VB "almost replacements" appears to be loads of one man efforts working in isolation.
The B4x project is good, and up and running, but is not VB compatible and I think the PC version is really a Java translator.
PureBasic is good, but never got enough users to be a full time job, again not VB compatible and a limited GUI, and no Objects.
twinBasic has now got closer than any other.
My own view is that with a bit more work on the Windows GUI I could start using it seriously.
VScode has its limitations, and its not like VB, but I quite like a lot of it.

**Comment by DinyaZ** [comment]
I think TwinBASIC and RAD Basic can be a winning combo. 
.
By the way, I have tried the B4x, a good project, but it hasnt too many similarities with the VB6, and of course it is not VB6 compatible. I thought they only use the VB6 "myth" as a marketing phrase. I think the best "VB6 replacement" try was the JABACO (https://www.jabaco.org/showcase.html) and it was already an almost useable tool in beta phase when its developer simply disappeared.

**Comment by uwbwsvd** [comment]
By the way:
Pascal: introduced as a teaching language by Niklaus Wirth at ETH Zurich.
Python: by Guido van Rossum at the Centre for Mathematics (Centrum voor Wiskunde en Informatica)
C: The Bell Laboratories (also Bell Labs, formerly Bell Telephone Laboratories, BTL for short) are the former research division 
Javacript: Netscape was founded on 4 April 1994 as Mosaic Communications Corporation by Marc Andreessen, one of the co-developers of the NCSA Mosaic web browser, and computer entrepreneur James H. Clark, who made the foundation possible in the first place with US$9 million.
Java: Sun Microsystems, usfounded out of Stanford University in 1982,

Ergo: Success is linked to university  or money.

Translated with www.DeepL.com/Translator (free version)

**Comment by Kr00l** [comment]
We can just wish Wayne great success and hope that it withstands until v1 and hope for a boom then.

Just few cents from me:
I would be careful with collaboration in regard to shares of the business. (If it's meant like that for the RAD basic guy) In long-term it complicates things and can get risky.

**Comment by dmrvb** [comment]
I don't think tB is going to be another VB6, or Java, and make a Wayne a multi billionaire, but with a bit of luck it could make him a very comfortable living and employ a small team.
I think B4x was/is quite successful.
Away from languages there are quite a few one-man projects that have succeeded.
Cooledit was doing well but was then sold to Adobe, and Help and Manual (a programming related project) has been going for years and supports a couple of people.
tB has a lot of potential.

**Comment by WaynePhillipsEA** [comment]
> What if you try to contact the RAD Basic developer guy? His goal is the same, to develop a modern, 100% VB6 compatible language. I think, your compiler and "under the hood" details are much more mature, but his project has also an important advantage: he develops a clear, simple, and user friendly GUI, editor, form designer interface, which is easier to use, and resembles to the VB6. I think, the biggest disadvantage of TwinBASIC now is the VSCode reliance.

I do appreciate your points, however we are already planning to move to Theia (discussed in #458), and this will allow us to take full control over the development environment.  Usability is set to improve massively once we make the move.

Having read many of Carles comments over at VBF, and from playing with the alpha version of RAD basic, I just can't see a collaboration being possible.  Carles is a nice guy, and whilst I don't wish to be disrespectful to him, the work he's produced so far does not give me much confidence that he is capable of producing a commercial grade compiler, debugger, etc.   Clearly his 100 patreons see things differently to me... so I wish him well.

With tB, so far, it's all about the compiler.  Probably 90%+ of my effort has so far gone into the compiler & debugger.  Once I do get around to working on the IDE side of things, you're going to see big improvements.  Stay tuned!

**Comment by KarlSevenSeven** [comment]
> With tB, so far, it's all about the compiler.

Yes, that's the VERY main part.
Along with language completeness and compatibility.

**Comment by dmrvb** [comment]
It's sad that the current income is not huge but that's not really surprising, no sensible person would give money to yet another one man band dreaming of re-creating VB6 (how do I insert a smiley in these posts??) .
Perhaps we need a small change of emphasis. Judging by the posting here a lot of effort is going into adding new language features and this should be left till later.
tB is not VB compatible because its GUI is very limited. Once it can start importing and compiling small (or even big) VB6 projects, including Forms, then things might start to get a bit more momentum. We probably don't even need 100% GUI compatibility as long as an upgrade wizard, or a days work, can get things running. Stuff like "This_Click handles This.Click" is really no great compatibility issue.
For me the recent WinForms looks like a HUGE step forward and with a few more controls and functionality I could start getting some of my things running in tB. Maybe we should have a place where we can put a few "written in tB" EXEs for people to have a look at?

**Comment by mansellan** [comment]
Great points @dmrvb.

I think we need a roadmap to v1. It's all well and good dreaming of future language improvements (guilty!), but what is actually needed for v1? For my money:

- Back-compatibility bugs addressed as number 1 priority (CHECK - has always been so)
- Built-in libraries replicated (MOSTLY CHECK - in progress) 
- VB6 forms complete, including built-in controls to (a minimum of) feature parity (in progress) 
- 3rd party OCX support (CHECK?)
- Win-x86 target (CHECK)

I believe that those are the MVP for v1. Bonus points for:

- Extra in-box VB6 controls (Winsock, MAPI, etc)
- Win-x64, fully prod-ready (CHECK?)
- Optimised compilation (It's pretty fast already...)
- VbWatchdog
- Inheritance (Via might already be optimal..._)
- Edit & Continue
- XPlat

I think it's entirely reasonable to focus on the v1 MVP; the point where marketing can begin in earnest and the product is suitable for customers to start building commercial applications (using VB6 source). All the "Beyond VB6" stuff can be in v2+, IMO.

tl;dr:  don't try to put all the ideas in v1.

**Reply by VB6-Programming** [reply]
I agree with @mansellan's list.
The major requirement is to be able to import an existing VB6 project with the expectation it will run with (almost) no changes.  And then upgrade it to 64bit (not necessarily optimised).

I would add that some form of Escrow should be implemented to ensure that the tB source code becomes available should Wayne not continue with the project. This would reassure anyone concerned about any 'one man band' issues.

I've got to add just how impressed I am with twinBASIC so far, well done Wayne. 👍

**Reply by WaynePhillipsEA** [reply]
Thanks!   Over the weekend I'm planning to tweak the roadmap.  We need a bit of a reshuffle of priorities to aim for a v1 release.

**Comment by bclothier** [comment]
FWIW: this seems more apropos as a discussion rather than an issue. Will convert accordingly.
